### PR TITLE
Company mode

### PR DESCRIPTION
**Pull request by rgrinberg**
@tuncer if you use company mode, would you mind giving this branch a try?

### CODE REVIEW

**Comment by rgrinberg** [comment]
OK I've made the company support completely optional to unbreak older utops.

**Comment by tuncer** [comment]
I hadn't configure utop.el yet. I can give it a try if you're still in need of basic testing. I don't think I would test advanced features if I did.

**Comment by rgrinberg** [comment]
I just thought that you wanted to give this feature a try since you asked about the original PR. I've tested this feature myself well enough so it's not necessary for you to do it as well unless you're comfortable with it.

**Comment by rgrinberg** [comment]
@diml I'm parsing `utop -version` to detect which utop we're dealing with. Not a thing of beauty, but it will do for now.

**Comment by diml** [comment]
What about saying that the first line utop prints when started with `utop -emacs` is `protocol-version:N`, and if it doesn't print it then the emacs mode assumes that it's using protocol version 1?

**Comment by rgrinberg** [comment]
OK sure, that's done.

File: src/top/utop.el [hunk 1]
```diff
@@ -635,10 +656,18 @@ it is started."
...
-         ;; Remove it and continue
-         (setq lines (cdr lines)))
+         ;; process first line. if we haven't tried reading the version, we'll
+         ;; trying to do it now.
+         (let ((line (car lines)))
+           (if utop--read-version
+               (utop-process-line line)
+             (progn
+               (save-match-data
+                 (when (string-match "protocol-version:\\([0-9]+\\)" line)
```
Comment (diml) [inline src/top/utop.el]: If the line doesn't match, do we still process it normally or does it end up being discarded?

Comment (rgrinberg) [reply src/top/utop.el]: Good point. I'll fix that.

**Comment by rgrinberg** [comment]
Will release 2.1.0 with this change.

**Pull request merged**